### PR TITLE
split

### PR DESCRIPTION
**Pull request by cr3mebrulee**
@ioannnikulin 
According to your concept keep rules "make" and "make fclean" for minishell and libft separately i added several rules to minishell makefile that we could clean .o of libft/tests. also i modified "make fclean", that it cleanes .o of minishell sources.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -70,6 +71,9 @@ testclean:
 testfclean: testclean
 	$(PREFIX)rm -f $(TEST_FNAME)
 
+test_libft_fclean:
```
Comment (ioannnikulin) [inline Makefile:74]: Let's keep uniform naming. Either this can be "pretestfclean", or rename all the others from "pre" to "libft". I prefer pre, since in theory we also can have other dependencies (like minilibx)

File: libft/sources/strings/ft_split_buf.c [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+/*   Updated: 2024/10/02 20:00:41 by taretiuk         ###   ########.fr       */
+/*                                                                            */
+/* ************************************************************************** */
+
+#include "../../sources/strings/strings_internal.h"
+#include "libft.h"
+
+int find_operator(const char *str, t_operator_array *op_array, size_t *op_len)
+{
+	for (size_t i = 0; i < op_array->count; i++)
```
Comment (ioannnikulin) [inline libft/sources/strings/ft_split_buf.c:18]: norme!

File: libft/sources/strings/strings.h [hunk 3]
```diff
@@ -3,16 +3,18 @@
...
 /*   Created: 2024/08/13 20:55:01 by inikulin          #+#    #+#             */
-/*   Updated: 2024/09/22 22:35:17 by inikulin         ###   ########.fr       */
+/*   Updated: 2024/10/02 11:36:13 by taretiuk         ###   ########.fr       */
 /*                                                                            */
 /* ************************************************************************** */
 
 #ifndef STRINGS_H
 # define STRINGS_H
 
+#include "../../sources/strings/strings_internal.h"
```
Comment (ioannnikulin) [inline libft/sources/strings/strings.h:16]: I see. So the user of your function DOES need the structure, it's a parameter. In this case you might as well include the whole content of that _internal.h here. The point of _internal.h is to NOT expose it to the user. If _internal.h is included here, and this file is included in libft.h, this means that the libft user sees declarations from _internal.h. Which makes it not _internal_ :)

File: libft/tests/strings/ft_split_buf_test.c [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+		{"mkdir new_dir ", "&&", "cd new_dir ", "&&", "touch file.txt", NULL},
+		{"mv new_dir old_dir", NULL}
+	};
+	for (int i = 0; i < NUM_TEST_CASES; i ++)
+	{
+		int sz = 0;
+		char **split_op = ft_split_buf(str_arr.strs[i].str, &op_arr, &sz);
+		for (int j = 0; j < sz; j ++)
+		{
+			if (split_op[j] == NULL || t[i][j] == NULL)
```
Comment (ioannnikulin) [inline libft/tests/strings/ft_split_buf_test.c:76]: I'm not exactly sure what you're doing here, but consider
`assert((split_op[j] == NULL) == (t[i][j] == NULL));`

File: libft/tests/strings/ft_split_buf_test.c [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+			{
+				printf("Error: NULL pointer detected in 2\n");
```
Comment (ioannnikulin) [inline libft/tests/strings/ft_split_buf_test.c:78]: what's the difference between this block and the next assert?

**Comment by cr3mebrulee** [comment]
@ioannnikulin 
1) my last PR was canceled, i handled the problem, but git doesn't see previously modified and commited files, how can i create commit them again and make new pull request?
2) CI sees the compilation Error from make fulltest which occur because of error in tests/memory. Will my next PR also fail because of it? Should i fix it?

File: libft/sources/strings/ft_split_buf.c [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+		cwi++;
+	}
+	return (0);
+}
+
+char	**ft_split_buf(const char *s, void *delim_ar, int *sz)
+{
+	char		**res;
+	int			token_q;
+	t_delims	*ar;
```
Comment (ioannnikulin) [inline libft/sources/strings/ft_split_buf.c:90]: Nope. This doesn't make sense. If the user of your function wants to call this function, he'd have to create an array of t_delims. This means that this structure should be not internal, its definition should be accessible from strings.h.

File: libft/tests/strings/ft_split_buf_test.c [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+		{"mkdir new_dir ", "&&", "cd new_dir ", "&&", "touch file.txt", NULL},
+		{"mv new_dir old_dir", NULL}
+	};
+	for (int i = 0; i < NUM_TEST_CASES; i ++)
+	{
+		int sz = 0;
+		char **split_op = ft_split_buf(str_arr.strs[i].str, &op_arr, &sz);
+		for (int j = 0; j < sz; j ++)
+		{
+			if (split_op[j] == NULL || t[i][j] == NULL)
```
Comment (ioannnikulin) [inline libft/tests/strings/ft_split_buf_test.c:77]: what's the difference between this block and the next assert?

File: libft/tests/strings/ft_split_buf_test.c [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+			{
+				ft_printf("Error: NULL pointer detected in 2\n");
+				continue;
+			}
+			assert((split_op[j] == NULL) == (t[i][j] == NULL));
+			if (split_op[j] == NULL)
+				break;
+			assert(strcmp(split_op[j], t[i][j]) == 0);
+		}
```
Comment (ioannnikulin) [inline libft/tests/strings/ft_split_buf_test.c:86]: please free the memory of current test

File: libft/tests/strings/ft_split_buf_test.c [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+	 char *t[NUM_TEST_CASES][MAX_ARGS] =
+	 {
+		{"cat ", "(", "file1.txt file2.txt ", ")", "|", "grep \"keyword\"", NULL},
+		{"cat ", "<", "input.txt ", "&&", "echo ", "|", "grep \"Done\"", NULL},
+		{"grep \"error\" log.txt ", "||", "echo \"No errors found\"", NULL},
+		{"gcc main.c -o program ", ">>", "build.log", NULL},
+		{"find . -name \"*.txt\" ", "|", "xargs rm -rf ", "||", "fer ", "&&", "def ", "&", "echo \"Hello\"", NULL},
+		{"echo \"Compiling...\" ", "&&", "gcc main.c -o program ", "&&", "./program", NULL},
+		{"mkdir new_dir ", "&&", "cd new_dir ", "&&", "touch file.txt", NULL},
+		{"mv new_dir old_dir", NULL}
```
Comment (ioannnikulin) [inline libft/tests/strings/ft_split_buf_test.c:69]: please add a one operator only test (like "&&"), an empty string test, multiple brackets test (like "()(()())"), incorrect brackets test (like ")(())(")

File: libft/tests/strings/ft_split_buf_test.c [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+			{
+				ft_printf("Error: NULL pointer detected in 2\n");
+				continue;
+			}
+			assert((split_op[j] == NULL) == (t[i][j] == NULL));
+			if (split_op[j] == NULL)
+				break;
+			assert(strcmp(split_op[j], t[i][j]) == 0);
+		}
+	}
```
Comment (ioannnikulin) [inline libft/tests/strings/ft_split_buf_test.c:87]: please free delims and strings array (or you could have defined them on the stack, without mallocs actually)

File: libft/sources/strings/strings_internal.h [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+} t_string;
+
+typedef struct s_string_array
+{
+	t_string	*strs;
+	size_t	count;
+} t_strings;
+
+
+char	**ft_split_buf(const char *s, void *delim_arr, int *sz);
```
Comment (ioannnikulin) [inline libft/sources/strings/strings_internal.h:34]: Isn't delim_arr also const?

**Comment by ioannnikulin** [comment]
@cr3mebrulee 1. checkout to that branch, push it (or maybe it's here in Github already), recreate the PR
2. we should merge only those PRs that pass the tests. This PR in its current state doesn't pass the tests, because you have an unused variable introduced in this PR. When you remove it (or use), the problem will go away, we'll merge the PR and you won't have to deal with this exact error in the next PRs. We still can merge other PRs, since this variable doesn't exist in them and won't break the tests there.

**Comment by ioannnikulin** [comment]
Also please rename the PR, I get confused every time I try to read the changes

File: libft/sources/strings/ft_split_buf.c (hunk 6 above)
Comment (cr3mebrulee) [reply libft/sources/strings/ft_split_buf.c:87]: @ioannnikulin please check if my solution to include internal header in strings.h make sense? then i include trings.h in all files.

File: libft/sources/strings/strings_internal.h (hunk 11 above)
Comment (cr3mebrulee) [reply libft/sources/strings/strings_internal.h:32]: well, i cast it to the t_delims * nside of the function. if it's ok, i can make prototype with const void *

File: libft/tests/strings/ft_split_buf_test.c (hunk 9 above)
Comment (cr3mebrulee) [reply libft/tests/strings/ft_split_buf_test.c:69]: done

File: libft/tests/strings/ft_split_buf_test.c (hunk 7 above)
Comment (cr3mebrulee) [reply libft/tests/strings/ft_split_buf_test.c:129]: The first assertion: assert((split_op[j] == NULL) == (t[i][j] == NULL)); checks that both pointers are either NULL or both are valid (non-NULL). If NULL has detected it break the loop preventing from comparison. Make sense?

File: libft/sources/strings/strings.h [hunk 12]
```diff
@@ -3,16 +3,18 @@
...
 /*   Created: 2024/08/13 20:55:01 by inikulin          #+#    #+#             */
-/*   Updated: 2024/09/22 22:35:17 by inikulin         ###   ########.fr       */
+/*   Updated: 2024/10/09 16:11:55 by taretiuk         ###   ########.fr       */
 /*                                                                            */
 /* ************************************************************************** */
 
 #ifndef STRINGS_H
 # define STRINGS_H
 
+#include "strings_internal.h"
```
Comment (ioannnikulin) [inline libft/sources/strings/strings.h:16]: This means all declarations from strings_internal.h are included into strings.h and hence into libft.h. This is undesired. I suggest including strings_internal only into ft_split_buf.c and ft_split_buf_utils.c. Including strings.h into strings_internal.h is fine, not the opposite.

File: libft/sources/strings/strings_internal.h [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+
+typedef struct s_string_array
+{
+	t_string	*strs;
+	int			error;
+	size_t		count;
+} t_strings;
+
+
+char	**ft_split_buf(const char *s, const void *delim_arr, int *sz);
```
Comment (ioannnikulin) [inline libft/sources/strings/strings_internal.h:34]: Why double declaration? Having it only in strings.h is enough. This way, if you decide to change anything later, you;d have to change this in two places.

File: libft/sources/strings/ft_split_buf.c (hunk 6 above)
Comment (ioannnikulin) [reply libft/sources/strings/ft_split_buf.c:87]: no, it breaks the idea of internal headers. see the comment in the code.

File: libft/tests/strings/ft_split_buf_test.c (hunk 7 above)
Comment (ioannnikulin) [reply libft/tests/strings/ft_split_buf_test.c:129]: I don't want to use things like
>> ft_printf("Error:
because it demands that I as a tester actually read through the whole testing result output. If this situation is an error, which means the function isn't functioning properly, I'd expect an `assert` call that stops everything. Then the error is more clearly visible.

File: libft/sources/strings/strings.h (hunk 12 above)
Comment (cr3mebrulee) [reply libft/sources/strings/strings.h:16]: Make sense, thanks

File: libft/sources/strings/strings_internal.h (hunk 13 above)
Comment (cr3mebrulee) [reply libft/sources/strings/strings_internal.h:34]: You are right

**Review by ioannnikulin — APPROVED** [review]
internal functions in test should have been static, apart from that looks good for now
